staging/src/k8s.io/apimachinery/pkg/util/net/interface.go

```diff
@@ -378,10 +378,10 @@ func chooseHostInterfaceFromRoute(routes []Route, nw networkInterfacer) (net.IP,
 }
 
 // If bind-address is usable, return it directly
-// If bind-address is not usable (unset, 0.0.0.0, or loopback), we will use the host's default
+// If bind-address is not usable (unset, 0.0.0.0), we will use the host's default
 // interface.
 func ChooseBindAddress(bindAddress net.IP) (net.IP, error) {
```

Comment:
this used to be called `ValidPublicAddrForMaster` which makes more sense: https://github.com/kubernetes/kubernetes/pull/17304/files#diff-6c6a7fefe37261e2fa8a010cf9e7ae08R519. 127.0.0.1 cannot be used as advertised external apiserver address. And as far as I see, it is still used in this sense, just misnamed.

Reply:
I guess the external apiserver address shows up in the `default/kubernetes` endpoints. And for network routing reason 127.0.0.1 does not make much sense there.

Reply:
what if you place client side load balancer which is listening to localhost, because localhost is the only IP available on all hosts?

Reply:
I don't think it's a good idea for the master to publish 127.0.0.1 as its discovery address

Reply:
> what if you place client side load balancer which is listening to localhost, because localhost is the only IP available on all hosts?
> 
> this is especially important for cases when kubernetes is installed on premises inside closed environment.

Note that 127.0.0.1 means something else for containers on that node as well (network namespaces...). Moreover, the `kubernetes` service and its endpoints are cluster global. Kubernetes clusters with only master nodes (where 127.0.0.1 might make sense) are not that common.

So as @liggitt I don't see much sense in 127.0.0.1 as discovery address.

Reply:
It seems you are missing the point of client side load-balancers for k8s HA installations, with multiple API servers.  
In case you have client side balancer, and run `kubeadm join`, api servers response with IP (e.g. 127.0.0.1) and kubelet as well as kube-proxy are using this configured by kubeadm ip to reach api servers. Without this approach, somebody need to create well known IP address instead of hostNetworking localhost. 
Example of the use case is in:
https://github.com/kubernetes/kubeadm/issues/487

Reply:
that requires *every* client of the apiserver to set up such a load balancer, which doesn't seem reasonable.

Reply:
@onorua as far as I see the function is only used to derive the default external address. You can still override it via `--advertise-address`, completely circumventing this function. So I propose to use that way for your very special setup.

Reply:
@sttts I would love to, but it is forbidden to use `--advertise-address` equals to  `127.0.0.1` in kubeadm, which is using that specific code to check which IP to use. That is why I've created this PR

Reply:
In that case, the function is not the right place to fix this. The function selects the external address from the bind address in the apiserver. And choosing the 127.0.0.1 for a bind address "localhost" is not correct in general.